pennylane/operation.py

```diff
@@ -709,6 +718,52 @@ def generator(self):
         """
         return [None, 1]
 
+    def parameter_frequencies(self):
+        r"""Returns the frequencies for each operator parameter with respect
+        to an expectation value of the form
+        :math:`\langle \psi | U(\mathbf{p})^\dagger \hat{O} U(\mathbf{p})|\psi\rangle`.
+        These frequencies encode the behaviour of the operator :math:`U(\mathbf{p})`
+        on the value of the expectation value as the parameters are modified.
+        For more details, please see the :mod:`.pennylane.fourier` module.
+
+        Returns:
+            list[tuple[int or float]]: Tuple of frequencies for each parameter.
+                Note that only non-negative frequency values are returned.
+
+        **Example**
+        >>> op = qml.CRot(0.4, 0.1, 0.3, wires=[0, 1])
+        >>> op.parameter_frequencies()
+        [(0.5, 1), (0.5, 1), (0.5, 1)]
+        For operators that define a generator, the parameter frequencies are directly
+        related to the eigenvalues of the generator:
+        >>> op = qml.ControlledPhaseShift(0.1, wires=[0, 1])
+        >>> op.parameter_frequencies()
+        [(1,)]
+        >>> gen_eigvals = tuple(np.linalg.eigvals(op.generator[0] * op.generator[1]))
+        >>> qml.gradients.eigvals_to_frequencies(gen_eigvals)
+        (tensor(1., requires_grad=True),)
+
+        For more details on this relationship, see :func:`.eigvals_to_frequencies`.
+        """
+        if self.num_params == 1:
+            # if the operator has a single parameter, we can query the
+            # generator, and if defined, use its eigenvalues.
+            op, coeff = self.generator
+            if op is None:
+                raise OperatorPropertyUndefined(
+                    f"Operation {self.name} does not have parameter frequencies."
+                )
+
+            if not isinstance(op, np.ndarray):
+                op = op.matrix
+            gen_eigvals = tuple(np.linalg.eigvals(op))
+            coeff = np.abs(coeff)
+            return tuple(coeff * val for val in qml.gradients.eigvals_to_frequencies(gen_eigvals))
```

Comment:
Would be very happy if you want to leave this if statement branch as a TODO, and simply raise `NotImplementedError` for now!

Reply:
Is it not okay like this? I found it nice that this allows to already understand the logical flow for default operations that do not define custom `parameter_frequencies`.

Reply:
Oh, this was more a comment that the generator will soon be breaking :) But definitely fine as is, no action needed here

Reply:
Ah, I see :) Yeah, I went for postponing any such trouble to later parts of this series of PR, just to get started :)